File: C++/ps_baekjoon/11505/main.cpp
#include <bits/stdc++.h>
using namespace std;
using ll = long long;

constexpr ll MOD = 1'000'000'007;
int n, m, k;

class SegTree
{
private:
    vector<ll> tree;
    const vector<ll>& data;
public:
    SegTree(const vector<ll>& data) :data{data} { tree.resize(data.size() * 4); }
    
    ll merge(ll left, ll right) { return (left * right) % MOD; }
    
    ll build(int node, int nodeLeft, int nodeRight)
    {
        if (nodeLeft == nodeRight)
            return tree[node] = data[nodeLeft];
        int mid = (nodeLeft + nodeRight) / 2;
        ll leftVal = build(node * 2, nodeLeft, mid);
        ll rightVal = build(node * 2 + 1, mid + 1, nodeRight);
        return tree[node] = merge(leftVal, rightVal);
    }
    
    ll query(int left, int right, int node, int nodeLeft, int nodeRight)
    {
        if (left <= nodeLeft && nodeRight <= right)
            return tree[node];
        if (right < nodeLeft || nodeRight < left)
            return 1;
        int mid = (nodeLeft + nodeRight) / 2;
        ll leftVal = query(left, right, node * 2, nodeLeft, mid);
        ll rightVal = query(left, right, node * 2 + 1, mid + 1, nodeRight);
        return merge(leftVal, rightVal);
    }
    
    ll update(int index, ll newVal, int node, int nodeLeft, int nodeRight)
    {
        if (index < nodeLeft || nodeRight < index)
            return tree[node];
        if (nodeLeft == nodeRight)
            return tree[node] = newVal;
        int mid = (nodeLeft + nodeRight) / 2;
        ll leftVal = update(index, newVal, node * 2, nodeLeft, mid);
        ll rightVal = update(index, newVal, node * 2 + 1, mid + 1, nodeRight);
        return tree[node] = merge(leftVal, rightVal);
    }
};

int main()
{
    cin.tie(0)->sync_with_stdio(0);
    cin >> n >> m >> k;
    vector<ll> arr(n + 1);
    for (int i = 1; i <= n; ++i)
        cin >> arr[i];
    SegTree segtree(arr);
    segtree.build(1, 1, n);
    for (int i = 0; i < m + k; ++i)
    {
        ll a, b, c;
        cin >> a >> b >> c;
        if (a == 1)
            segtree.update(b, c, 1, 1, n);
        if (a == 2)
            cout << segtree.query(b, c, 1, 1, n) << '\n';
    }
    return 0;
}

File: Python/ps_baekjoon/11505/main.py
import sys

readline, write = (sys.stdin.readline, sys.stdout.write)
res = []
MOD = 1000000007


class SegTree:
    def __init__(self, data):
        self.data = data
        self.tree = [0] * (len(data) * 4)

    def merge(self, left, right):
        return (left * right) % MOD

    def build(self, node, nodeLeft, nodeRight):
        if nodeLeft == nodeRight:
            self.tree[node] = self.data[nodeLeft]
            return self.tree[node]
        mid = (nodeLeft + nodeRight) // 2
        leftVal = self.build(node * 2, nodeLeft, mid)
        rightVal = self.build(node * 2 + 1, mid + 1, nodeRight)
        self.tree[node] = self.merge(leftVal, rightVal)
        return self.tree[node]

    def query(self, left, right, node, nodeLeft, nodeRight):
        if right < nodeLeft or nodeRight < left:
            return 1
        if left <= nodeLeft and nodeRight <= right:
            return self.tree[node]
        mid = (nodeLeft + nodeRight) // 2
        leftVal = self.query(left, right, node * 2, nodeLeft, mid)
        rightVal = self.query(left, right, node * 2 + 1, mid + 1, nodeRight)
        return self.merge(leftVal, rightVal)

    def update(self, index, newVal, node, nodeLeft, nodeRight):
        if index < nodeLeft or nodeRight < index:
            return self.tree[node]
        if nodeLeft == nodeRight:
            self.tree[node] = newVal
            return self.tree[node]
        mid = (nodeLeft + nodeRight) // 2
        leftVal = self.update(index, newVal, node * 2, nodeLeft, mid)
        rightVal = self.update(index, newVal, node * 2 + 1, mid + 1, nodeRight)
        self.tree[node] = self.merge(leftVal, rightVal)
        return self.tree[node]


n, m, k = [int(i) for i in readline().split()]
data = [0] + [[int(i) for i in readline().split()][0] for _ in range(n)]
st = SegTree(data)
st.build(1, 1, n)
for _ in range(m + k):
    a, b, c = [int(i) for i in readline().split()]
    if a == 1:
        st.update(b, c, 1, 1, n)
    if a == 2:
        res.append(str(st.query(b, c, 1, 1, n)))
write("\n".join(res))